content/2019-06-11_co-dela-server-zpravy-context-manager-a-jak-porovnavat-hodnoty-v-pythonu.md

```diff
@@ -0,0 +1,220 @@
+Title: Co dělá server: zprávy, context manager a jak porovnávat hodnoty v Pythonu
+Date: 2019-06-11 18:00:00
+Modified: 2019-06-11 20:00:00
+Author: Karolina Surma
+
+## is vs ==
+
+`==` je porovnání, které ověří hodnotu objektů, které jsou porovnávány.
+
+```python
+>>> 1.0 == 1
+True
+>>> 1.0 == True
+True
+```
+
+U každého datového typu můžeme upravit chování porovnávání přepsáním speciální metody `__eq__`. Psaly jsme o tom post [zde](https://roboprojekt.pyladies.cz/dvacaty-sraz-testy).
+
+A `is` vyhodnotí, jestli je to tentýž, stejný objekt.
+Příklad níže se vyhodnotí jako nepravda, protože dva prázdné seznamy jsou dva různé objekty.
+
+```python
+>>> [] is []
+False
+```
+
+Dva prázdné seznamy se sice rovnají, pokud se vyhodnotíme pomocí `==`, ale v budoucnosti nemusí tomu tak být:
+
+```python
+>>> a = []
+>>> b = []
+>>> a == b
+True
+>>> a.append(1)
+>>> a == b
+False
+```
+
+Zatímco `is` kontroluje, že je to úplně přesně to samé. 
+
+```python
+>>> 1.0 is 1
+False
+>>> 1 is 1 
+True
+```
+
+U posledního příkladu vzniká problém, protože u celých čísel se Python chová trochu nestandardně: nezáleží mu, jestli jde o tentýž objekt, nebo jiný objekt s hodnotou čísla 1.
+Navíc malá a velká čísla se chovají trochu jinak:
+
+```python
+>>> a = 1
+>>> b = 1
+>>> a is b
+True
+
+>>> a = 1000
+>>> b = 1000
+>>> a is b
+False
+```
+
+U některých objektů proto `is` může být zrádné. U `True`, `False` a `None` tento problém není, protože nemůžeme vytvořit další objekt "True", vždy se jedná o stejnou instanci logické hodnoty.
+Když chceme použit vyhodnocení atributu s _bool_ hodnotou v podmínce, neměly bychom ale používat ani `==`, ani `is` - správný způsob, jak takovou podmínku napsat, je:
+
+```python
+if atribut: # místo "if atribut is True"
+    print("Ha!")
+if not atribut: # místo "if atribut is False"
+    print("Nene!")
+```
+
+U `None` věc vypadá trochu jinak, zde opravdu chceme zjišťovat, jestli hodnotou atributu je `None`, nebo ne, takže:
+
+```python
+if atribut is None:
+    print("Jsem prázdný")
+if atribut is not None:
+    print("Nejsem prázdný")
+```
+
+`is` je trochu rychlejší a vyhodnocení podmínky tak trvá o něco méně, než v případě tradičního porovnávání.
+Jako zajímavost jsme se dozvěděly, že o `is` a `==` se točí v Pythonním světě diskuze, a dokonce autor Pythonu, Guido, se vyjádřil pro sjednocení porovnávání ve prospěch `==`. 
```

Comment:
Tohle by bylo dobré ocitovat, ale nemůžu dohledat zdroj. Nerad bychšířil fámy. Můžeme zůstat u těch diskuzí?
Příklad diskuze: https://discuss.python.org/t/demoting-the-is-operator-to-avoid-an-identity-crisis/86
(Guido už aktuálně pro sjednocení není; rozbilo by to moc existujícího kódu.)

```suggestion
Jako zajímavost jsme se dozvěděly, že o `is` a `==` se točí v Pythonním světě diskuze.
```